inspirehep/records/marshmallow/literature/common/external_system_identifier.py

```diff
@@ -0,0 +1,106 @@
+        "kekscan": "KEK scanned document",
+        "msnet": "AMS MathSciNet",
+        "osti": "OSTI Information Bridge Server",
+        "zblatt": "zbMATH",
+    }
+
+    schema_to_url_link_prefix_map = {
+        "ads": "http://adsabs.harvard.edu/abs/",
+        "cds": "http://cds.cern.ch/record/",
+        "euclid": "http://projecteuclid.org/",
+        "hal": "https://hal.archives-ouvertes.fr/",
+        "kekscan": "https://lib-extopc.kek.jp/preprints/PDF/",
+        "msnet": "http://www.ams.org/mathscinet-getitem?mr=",
+        "osti": "https://www.osti.gov/scitech/biblio/",
+        "zblatt": "http://www.zentralblatt-math.org/zmath/en/search/?an=",
+    }
+
+    def get_url_name(self, data):
+        schema = data.get("schema").lower()
+        url_name = self.schema_to_url_name_map.get(schema)
+        if url_name is None:
+            return missing
+        return url_name
+
+    def get_url_link(self, data):
+        schema = data.get("schema").lower()
+        value = data.get("value")
+        url_prefix = self.schema_to_url_link_prefix_map.get(schema)
+        if url_prefix is None:
+            return missing
+        if schema == "kekscan":
+            kekscan_link = self.get_link_for_kekscan_schema(value)
+            if kekscan_link is None:
+                return missing
+            return kekscan_link
+        return "{}{}".format(url_prefix, value)
+
+    def get_link_for_kekscan_schema(self, external_system_id_value):
+        extid = external_system_id_value.replace("-", "")
+        if (
+            len(extid) == 7
+            and not extid.startswith("19")
+            and not extid.startswith("20")
+        ):
+            year = "19" + extid[:2]
+        elif len(extid) == 9:
+            year = extid[:4]
+            extid = extid[2:]
+        else:
+            return None
+        yymm = extid[:4]
+        kekscan_url_prefix = self.schema_to_url_link_prefix_map["kekscan"]
+        return "{}{}/{}/{}.pdf".format(kekscan_url_prefix, year, yymm, extid)
+
+    @post_dump(pass_many=True)
+    def filter(self, data, many):
+        if not many:
+            if self.is_missing_url_name_or_link(data):
+                return {}
+            return data
+        data = self.take_ids_that_have_all_fields(data)
+        data = self.take_first_id_foreach_url_name(data)
+        return data
+
+    def take_ids_that_have_all_fields(self, external_system_ids):
+        return [
+            extid
+            for extid in external_system_ids
+            if not self.is_missing_url_name_or_link(extid)
+        ]
+
+    def is_missing_url_name_or_link(self, external_system_id):
+        return (
+            external_system_id.get("url_link") is None
+            or external_system_id.get("url_name") is None
+        )
+
+    def take_first_id_foreach_url_name(self, external_system_ids):
+        taken_url_names = set()
+        unique_ids = []
+        for external_system_id in external_system_ids:
```

Comment:
```python
unique_ids = set()
for external_system_id in external_system_ids:
    url_name = external_system_id.get("url_name")
    unique_ids.add(url_name)
return unique_ids
```

Reply:
the tests are failing with this change, I'll leave as is for now